include/swift/AST/ASTContext.h

```diff
@@ -595,6 +596,17 @@ class ASTContext final {
   getClangFunctionType(ArrayRef<AnyFunctionType::Param> params, Type resultTy,
                        FunctionTypeRepresentation trueRep);
 
+  /// Given a FunctionTemplateDecl and SubstituionMap, lookup and convert all
+  /// template parameters.
+  ///
+  /// \param func The FunctionTemplateDecl who's parameters are getting looked
+  /// up and converted. \param subst The SubstituionMap to lookup the template
+  /// parameters in. \param convertedParams A SmallVector to accumulate
+  /// substituted and coverted types as TemplateArgument in.
+  void getSubstitutedClangTypesForTemplateFunction(
+      clang::FunctionTemplateDecl *func, SubstitutionMap subst,
+      SmallVectorImpl<clang::TemplateArgument> &convertedParams);
```

Comment:
There is a bunch of duplication in the doc comment, and I'm not sure why the `ForTemplateFunction` suffix is present... Is the behavior going to be different when instantiating class template arguments?

How about something like (haven't checked for line length):

```
/// Translate Swift generic arguments to template arguments.
///
/// Returns an error if any parameter in \param templateParams is a non-type template
/// parameter or if there is a conversion failure for one of the generic arguments. 
bool getClangTemplateArguments(
  const TemplateParameterList *templateParams, SubstitutionMap genericArgs,
  SmallVectorImpl<clang::TemplateArgument> &templateArgs)
```

The type of the method + parameter names kinda' explain what the function is doing.

(The `bool` is a placeholder for an error type. You could replace it with `void` temporarily with a TODO.)

Note that there is a distinction between parameters and arguments, so naming the out parameter `SmallVectorImpl<clang::TemplateArgument> &convertedParams` is not right, since we are returning the template arguments, as reflected in the type.

Reply:
Thanks, that's good to know. Done.